cid/drainer.py

```diff
@@ -0,0 +1,168 @@
+           :param cluster: ECS cluster of which ECS container instance of interest is a member
+           :param container_instance_arn: ECS container instance ARN
+        """
+        logger = self.logger
+
+        if self.cache.get('InstanceIsDraining'):
+            logger.info("Container instance: %s previously set to draining", container_instance_arn)
+            return
+
+        logger.info(
+            "Setting status of container instance: %s to DRAINING...", container_instance_arn)
+
+        self.ecs_client.update_container_instances_state(
+            cluster=cluster, containerInstances=[container_instance_arn], status='DRAINING',
+        )
+
+        self.cache['InstanceIsDraining'] = True
+
+    def get_running_tasks(self, cluster, container_instance_arn):
+        """Return a list of tasks running on a container instance.
+           :param cluster: ECS cluster of which ECS container instance of interest is a member
+           :param container_instance_arn: ECS container instance ARN
+        """
+        logger = self.logger
+
+        list_tasks_response = self.ecs_client.list_tasks(
+                cluster=cluster, containerInstance=container_instance_arn)
+
+        logger.debug("Task list for container instance: %s: %s",
+                     container_instance_arn, list_tasks_response['taskArns'])
+
+        running_tasks = list_tasks_response['taskArns']
+
+        logger.debug("Container instance: %s has tasks: %s", container_instance_arn, running_tasks)
+
+        return running_tasks
+
+    def run(self):
+        logger = self.logger
+        message = self.message
+
+        logger.info("Lambda received the event %s", self.event)
+
+        asg_name = message['AutoScalingGroupName']
+        topic_arn = self.event['Records'][0]['Sns']['TopicArn']
+
+        if not ('LifecycleTransition' in message and
+                message['LifecycleTransition'] == 'autoscaling:EC2_INSTANCE_TERMINATING'):
+            logger.info("Ignoring non-EC2_INSTANCE_TERMINATING message")
+            return
+
+        ec2_instance_id = message['EC2InstanceId']
+        logger.info("Received termination notification for EC2 instance-id: %s, ASG: %s",
+                    ec2_instance_id, asg_name)
+
+        cluster, container_instance_arn = self.get_ecs_details(ec2_instance_id)
+        logger.info(
+            "EC2 instance-id: %s appears to be container instance ARN: %s, belonging to ECS cluster: %s",
+            ec2_instance_id, container_instance_arn, cluster
+        )
+
+        lifecycle_hook_name = message['LifecycleHookName']
+        logger.debug("Lifecycle hook name is %s", lifecycle_hook_name)
+
+        self.set_draining(cluster, container_instance_arn)
+
+        logger.info("Checking for running tasks...")
+        tasks_running = self.get_running_tasks(cluster, container_instance_arn)
+        logger.debug("Running tasks: %s", tasks_running)
+
+        if len(tasks_running) > 0:
```

Comment:
what if tasks refuse to stop or take too long?

Reply:
The hook is configured with a timeout and a fallback action, eg: 
```
  ASGTerminateHook:
    Type: AWS::AutoScaling::LifecycleHook
    Properties:
      AutoScalingGroupName: !Ref ECSAutoScalingGroup
      DefaultResult: ABANDON
      HeartbeatTimeout: 900
      LifecycleTransition: autoscaling:EC2_INSTANCE_TERMINATING
      NotificationTargetARN:
        Fn::ImportValue: !Sub ${CidStackName}-SNSTopicArn
      RoleARN: !GetAtt SNSLambdaRole.Arn
```

Reply:
https://docs.aws.amazon.com/autoscaling/latest/userguide/lifecycle-hooks.html#lifecycle-hook-results

Reply:
Yea, this is from AWS side, right? But when the lambda starts invoking itself via SNS how does it know if the hook has not timed out yet?

Reply:
Ah ok yes I see. If you try to complete a lifecycle action that's already been completed that's an error, so the code as-is should DTRT. That said, would be good to handle this condition explicitly.